pkg/sql/lease.go

```diff
@@ -573,20 +584,51 @@ func (t *tableState) acquire(
 ) (*tableVersionState, error) {
 	t.mu.Lock()
 	defer t.mu.Unlock()
-	// Wait for any existing lease acquisition.
+
+	// Block and wait for the new lease to be acquired on another routine.
 	t.acquireWait()
 
-	// Acquire a lease if no lease exists or if the latest lease is
-	// about to expire.
+	// Ensure a lease is acquired. Even though we finished waiting, another
+	// routine which finished acquiring a lease could have released it by now.
 	if s := t.mu.active.findNewest(); s == nil || s.hasExpired(timestamp) {
 		if err := t.acquireNodeLease(ctx, m, hlc.Timestamp{}); err != nil {
 			return nil, err
 		}
 	}
 
+	if t.mu.timer == nil {
+		t.startRefreshingRoutine(ctx, m)
+	}
+
 	return t.findForTimestamp(ctx, timestamp, m)
 }
 
+// startRefreshingRoutine begins the goroutine responsible for refreshing a
+// table lease.
+// TODO(joey): Currently this routine will run forever for each table
+// descriptor. Ideally, we stop the routine if the table descriptor hasn't been
+// used recently.
+//
+// t.mu needs to be locked.
+func (t *tableState) startRefreshingRoutine(ctx context.Context, m *LeaseManager) {
+	t.mu.timer = time.NewTimer(LeaseRenewalTimeout)
+	t.stopper.RunWorker(ctx, func(ctx context.Context) {
+		for {
+			select {
+			case <-t.mu.timer.C:
+				t.mu.Lock()
+				err := t.acquireNodeLease(ctx, m, hlc.Timestamp{})
```

Comment:
I think we need to add `acquireWait` here because I believe there's a race condition here. This routine can barge into `acquireNodeLease` while another routine in inside `acquireNodeLease` and just unlocked `t.mu` in order to call `m.LeaseStore.acquire`. This specifically happens because this routine doesn't respect the `acquiring` channel which acts as an implicit ordered lock.

(calling this now as I just thought of it and I'm fairly confident about it, albeit it's a rare condition, going to make the change)

Reply:
I added `acquireWait`, but I didn't end up making the change due to running into unknown deadlock problems.